### ISSUE TITLE
Introduce standardized exception for  an integrity constraint

### ISSUE THREAD

**Issue by mkouba** [eclipse-vertx/vertx-sql-client#1121]
#### Read me

Read this first before creating an issue:

- do not use this issue tracker to ask questions, instead use one of these [channels](https://vertx.io/community/). Questions will likely be closed without notice.
- you shall create a feature request only when it is general purpose enough.
- make sure that the feature is not already

#### Describe the feature

Give the simplest and best explanation.

#### Use cases

A list of use cases this feature will enable and the value it creates.

#### Contribution

Who should implement this feature ? are you volunteering for implementing this feature or
do you know that is able and willing implement this feature ?

**Comment by DavideD** [comment]
> so that higher layers can detect such a violation in a portable way

Hibernate Reactive for example

**Comment by vietj** [comment]
I believe that's not something the client can easily because it really depends on each database and implementation.

Looking at the code of JDBC client for PostgreSQL, it seems that `SQLIntegrityConstraintViolationException` is not supported.

@DavideD is `SQLIntegrityConstraintViolationException` used in practice in blocking Hibernate ? because if it does, it means it would not reliably work with PostgreSQL

**Comment by DavideD** [comment]
No, it's not a blocking issue. It just sounds like a nice idea.

**Comment by DavideD** [comment]
> @DavideD is `SQLIntegrityConstraintViolationException` used in practice in blocking Hibernate ? because if it does, it means it would not reliably work with PostgreSQL

I think Hibernate ORM tries different things to recongnized the exception. After a very quick look I can see it checks the error code from a `SQLException` or if the type of the exception is an `SQLIntegrityConstraintViolationException`. And it also uses dialect to have different approaches for different databases.

**Comment by DavideD** [comment]
Now that I'm having a better look at this, I'm noticing that PgException actually has several fields with info about the error. Maybe that's enough for us to figure out that's a constraint violation (or something else).

**Comment by mkouba** [comment]
> Now that I'm having a better look at this, I'm noticing that PgException actually has several fields with info about the error. Maybe that's enough for us to figure out that's a constraint violation (or something else).

Yes, there seems to be a [bunch of error codes for constraint violations](https://www.postgresql.org/docs/current/errcodes-appendix.html) for PostgreSQL; and the `PgException.getCode()` could be used to access the code.

I'm going to close this issue.
